atef/tools.py

```diff
@@ -0,0 +1,295 @@
+    _unresponsive_time: ClassVar[float] = 100.0
+
+    @staticmethod
+    def _result_from_output(host: str, output: str) -> PingResult:
+        """
+        Fill a PingResult from the results of the ping program.
+
+        Parameters
+        ----------
+        host : str
+            The hostname that ``ping`` was called with.
+        output : str
+            The decoded output of the subprocess call.
+
+        Returns
+        -------
+        PingResult
+        """
+        # NOTE: lazily ignoring non-millisecond-level results here; 1 second+
+        # is the same as non-responsive if you ask me...
+        times = [float(ms) for ms in Ping._time_re.findall(output)]
+
+        if not times:
+            return PingResult(
+                result=Result(severity=Severity.error),
+                alive=[],
+                unresponsive=[host],
+                min_time=Ping._unresponsive_time,
+                max_time=Ping._unresponsive_time,
+                times={host: Ping._unresponsive_time},
+            )
+
+        return PingResult(
+            result=Result(severity=Severity.success),
+            alive=[host],
+            unresponsive=[],
+            min_time=min(times),
+            max_time=max(times),
+            times={host: sum(times) / len(times)},
+        )
+
+    async def ping(self, host: str) -> PingResult:
+        """
+        Ping the given host.
+
+        Parameters
+        ----------
+        host : str
+            The host to ping.
+
+        Returns
+        -------
+        PingResult
+        """
+
+        if sys.platform == "win32":
+            args = ("/n", str(self.count))
+        else:
+            args = ("-c", str(self.count))
+
+        proc = await asyncio.create_subprocess_exec(
+            str(shutil.which("ping")),
+            *args,
+            host,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.DEVNULL,
+        )
+        assert proc.stdout is not None
+        output = await proc.stdout.read()
+        await proc.wait()
+        return self._result_from_output(host, output.decode(self.encoding))
+
+    async def run(self) -> PingResult:
+        """
+        Run the "Ping" tool with the current settings.
+
+        Returns
+        -------
+        PingResult
+        """
+        result = PingResult(
+            result=Result(),
+            unresponsive=[],
+            alive=[],
+            times={},
+            min_time=0.0,
+            max_time=0.0,
+        )
+        self.result = result
+
+        if not self.hosts:
+            return result
+
+        ping_by_host: Dict[str, Union[Exception, PingResult]] = {}
+
+        async def _ping(host: str) -> None:
+            try:
+                ping_by_host[host] = await self.ping(host)
+            except Exception as ex:
+                ping_by_host[host] = ex
+
+        tasks = [asyncio.create_task(_ping(host)) for host in self.hosts]
```

Comment:
Is this line and the implications of doing this in general the reason why you async-ified the checking process?

Reply:
I think that task parallelization (be it for the tools, comparisons, or anything else in atef) will be important for design decisions and of course performance.  I realized that adding threads would be going in the direction of the "old" ophyd and would create trouble down the line for us (with v2 being dangled in front of us).

It's not too late to switch back to threaded, if you feel that it's premature to go the asyncio route.

Reply:
The only problem with the async approach is the one we outlined above: ophyd/pyepics do not (cannot) support it easily, e.g. any aync ophyd method is probably just a threaded method in disguise

Reply:
Let's think about both scenarios a bit - my reply above was not well-thought out enough. Enough so to probably be misleading.

Any big things I might be missing here?

### Threading
* We want to parallelize general tasks such as the tools provided in this PR, so we spawn threads and wait on some completion `threading.Event` or utilize a `ThreadPool`
   1. These are ugly and tougher to maintain/understand, but they work
   2. Propagation of exceptions/errors gets a bit more complicated, but it's doable
* We want `ophyd.EpicsSignal.get()` and other blocking ophyd stuff to be done in parallel
   1. As in the general task parallelization, we similarly put such calls into background threads
   2. `.wait_for_connection()` and `.get()` occasionally yield control due to their usage of `time.sleep()`, at which point other threads can run
* Data reduction/subscription handling is done via callback with the acquisition thread blocking (no changes really necessary here)

### asyncio
* Task parallelization as in this PR is done rather tersely with `create_task`/`wait`
* We want to somehow take non-async ophyd stuff and integrate it with asyncio
   1. We pass `.get()` calls off to run in the asyncio event loop's [executor thread pool](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)
   2. Assuming ophyd's own threading doesn't get in the way, it should be as simple as `value = await loop.run_in_executor(None, signal.get)`
   4. `.wait_for_connection()` and `.get()` will occasionally yield control due to `time.sleep()`, at which point other threads (including our asyncio Event loop thread) can run
* We'll want to move subscription/data reduction processing to asyncio (looks like I already [started that](https://github.com/klauer/atef/blob/c2d65ce16a21f6d5463f3682baad8c1800540ee7/atef/ophyd_helpers.py#L217))

##### My take

I think that even though the `.get()` replacement is "a threaded method in disguise", if we want to be asyncio-friendly in our development, it'd probably be easier to build it in now rather than later. It will certainly not be perfect, but I expect it could run well enough for what we need - and the code will probably be cleaner.

But you shouldn't just take my word on this; I'll dig a bit deeper into the above and we can revisit.

Reply:
I need to brush up on how asyncio works. I didn't know about `run_in_executor` for example. I didn't necessarily mean to apply that a "threaded method in disguise" is necessarily a very bad thing- it just seems to be the only reason why we wouldn't go the async route.

Reply:
Implications of the question aside - I think this is a really good thing to discuss. (Whether a line comment in a PR is the right spot for it or not, meh...)

That is to say:
* We definitely don't want to flippantly decide on big architectural things like this
* I should be able to solidly justify why I think we should just go asyncio now rather than later
* The implications of threaded/asyncio should be clear before we commit

Reply:
I'm at the same point as before where it intuitively seems right to me to go to async to help the checker run at scale, but I would have trouble verbalizing an argument for it.

I think we're going from single threaded to async here and not doing a threaded implementation at all which seems more correct than doing it the other way to me. The main advantage of async over threaded is brevity and stricter control over the execution compared to letting the python thread scheduler take the wheel, right? Theoretically fewer race conditions?

Reply:
I think the parallel data cache implementation shows that this can be cleaner and viable even now with ophyd as-is. I'm reasonably convinced and hope @ZLLentz will be, too

(More on this later, it's lunchtime)